third_party/scsi_mock/engine.cc

```diff
@@ -71,6 +78,13 @@ ScsiToNvmeResponse ScsiToNvme(unsigned char* cmd_buf, unsigned short cmd_len,
   translator::CompleteResponse cpl_resp =
       translation.Complete(nvme_cpl, buffer_in, sense_buffer);
 
+  if (cpl_resp.status == ApiStatus::kFailure) {
+    Print("Incorrect usage of Translation Library API");
+    ScsiToNvmeResponse resp = {.return_code = 0x40, .alloc_len = 0};
+    translation.AbortPipeline();
```

Comment:
Isn't this already called by `Complete`?

Reply:
When `Complete` returns an `ApiStatus::kFailure`, it doesn't call `AbortPipeline`. I'm still investigating whether it is necessary to call `AbortPipeline`. In any case, the user (`engine.cc`) probably shouldn't be responsible for calling `AportPipeline`

Reply:
Ah, I see e.g. https://github.com/googleinterns/scsi2nvme/blob/master/lib/translator/translation.cc#L153-L160 could cause a failure return with allocated resources. I agree this should probably be changed to clean up automatically.